File: src/api/quiz.ts
import { authFetch, apiUrl } from "@/lib/authFetch";

// ── TypeScript interfaces matching real API response shapes ──

export interface QuizQuestion {
  id: string;
  question_index: number;
  question_text: string;
  question_type: "mcq" | "true_false" | "fill_in_the_blank";
  difficulty: "easy" | "medium" | "hard";
  options: string[] | null;
  source_page: number | null;
}

export interface Quiz {
  id: string;
  pdf_id: string;
  pdf_ids: string[];
  status: string;
  question_count: number;
  questions: QuizQuestion[];
  title: string | null;
  chat_session_id: string | null;
  has_result: boolean;
  score: number | null;
  total: number | null;
  percentage: number | null;
  grade: string | null;
  created_at: string;
}

export interface PerQuestionResult {
  question_id: string;
  question_text: string;
  is_correct: boolean;
  user_answer: string | null;
  correct_answer: string;
  explanation: string | null;
}

export interface QuizSubmitResult {
  session_id: string;
  score: number;
  total: number;
  percentage: number;
  grade: string;
  per_question: PerQuestionResult[];
  weak_topics: string[];
  recommendation: string;
}

export interface AppendResult {
  id: string;
  question_count: number;
  new_questions_added: number;
  questions: QuizQuestion[];
}

export interface GenerateConfig {
  count: number;
  question_type: "mcq" | "true_false" | "fill_in_the_blank";
  difficulty: "easy" | "medium" | "hard";
  topic?: string | null;
}

// ── Helper to handle API errors consistently ──

async function handleResponse<T>(res: Response): Promise<T> {
  if (res.ok) return res.json();

  if (res.status === 409) throw new Error("PDF_NOT_READY");
  if (res.status === 404) throw new Error("NOT_FOUND");
  if (res.status === 422) {
    const body = await res.json().catch(() => null);
    throw new Error(body?.detail || "Validation error — check request body");
  }
  throw new Error(`Server error (${res.status})`);
}

// ── API functions ──

/** POST /api/quiz/{pdf_id}/generate */
export const generateQuiz = async (
  primaryPdfId: string,
  config: GenerateConfig,
  opts?: {
    pdf_ids?: string[];
    chat_session_id?: string | null;
    title?: string | null;
  },
): Promise<Quiz> => {
  const res = await authFetch(apiUrl(`/api/quiz/${primaryPdfId}/generate`), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      count: config.count,
      question_type: config.question_type,
      difficulty: config.difficulty,
      topic: config.topic ?? null,
      pdf_ids: opts?.pdf_ids ?? [],
      chat_session_id: opts?.chat_session_id ?? null,
      title: opts?.title ?? null,
    }),
  });
  return handleResponse<Quiz>(res);
};

/** POST /api/quiz/{quiz_id}/retake — clone with same questions */
export const retakeQuizSame = async (
  quizId: string,
  opts?: { chat_session_id?: string | null },
): Promise<Quiz> => {
  const res = await authFetch(apiUrl(`/api/quiz/${quizId}/retake`), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ chat_session_id: opts?.chat_session_id ?? null }),
  });
  return handleResponse<Quiz>(res);
};

/** GET /api/quiz/chat-session/{chat_session_id}/cards */
export const listQuizCardsForSession = async (chatSessionId: string): Promise<Quiz[]> => {
  const res = await authFetch(apiUrl(`/api/quiz/chat-session/${chatSessionId}/cards`));
  return handleResponse<Quiz[]>(res);
};

/** GET /api/quiz/{pdf_id}/list */
export const listQuizzes = async (pdfId: string): Promise<Quiz[]> => {
  const res = await authFetch(apiUrl(`/api/quiz/${pdfId}/list`));
  return handleResponse<Quiz[]>(res);
};

/** GET /api/quiz/{quiz_id} — no answers, for active quiz UI */
export const getQuiz = async (quizId: string): Promise<Quiz> => {
  const res = await authFetch(apiUrl(`/api/quiz/${quizId}`));
  return handleResponse<Quiz>(res);
};

/** POST /api/quiz/{quiz_id}/append */
export const appendQuestions = async (
  quizId: string,
  opts: {
    count?: number;
    question_type?: string | null;
    difficulty?: string | null;
  },
): Promise<AppendResult> => {
  const res = await authFetch(apiUrl(`/api/quiz/${quizId}/append`), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(opts),
  });
  return handleResponse<AppendResult>(res);
};

/** POST /api/quiz/{quiz_id}/submit */
export const submitQuiz = async (
  quizId: string,
  answers: Record<string, string>,
): Promise<QuizSubmitResult> => {
  const res = await authFetch(apiUrl(`/api/quiz/${quizId}/submit`), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ answers }),
  });
  return handleResponse<QuizSubmitResult>(res);
};

/** GET /api/quiz/{quiz_id}/result */
export const getQuizResult = async (
  quizId: string,
): Promise<QuizSubmitResult> => {
  const res = await authFetch(apiUrl(`/api/quiz/${quizId}/result`));
  return handleResponse<QuizSubmitResult>(res);
};

/** DELETE /api/quiz/{quiz_id} */
export const deleteQuiz = async (
  quizId: string,
): Promise<{ success: boolean; message: string }> => {
  const res = await authFetch(apiUrl(`/api/quiz/${quizId}`), {
    method: "DELETE",
  });
  return handleResponse<{ success: boolean; message: string }>(res);
};

File: server/app/models/quiz.py
import uuid
from datetime import datetime
from sqlalchemy import String, Integer, Float, Boolean, Text, ForeignKey, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.types import DateTime
from app.core.database import Base


class QuizSession(Base):
    __tablename__ = "quiz_sessions"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    pdf_id: Mapped[str] = mapped_column(String, ForeignKey("pdfs.id", ondelete="CASCADE"), nullable=False)
    user_id: Mapped[str] = mapped_column(String, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="active")
    question_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    pdf: Mapped["PDF"] = relationship("PDF", back_populates="quiz_sessions")
    user: Mapped["User"] = relationship("User", back_populates="quiz_sessions")
    questions: Mapped[list["QuizQuestion"]] = relationship("QuizQuestion", back_populates="session", cascade="all, delete-orphan")
    answers: Mapped[list["QuizAnswer"]] = relationship("QuizAnswer", back_populates="session", cascade="all, delete-orphan")
    result: Mapped["QuizResult | None"] = relationship("QuizResult", back_populates="session", uselist=False, cascade="all, delete-orphan")


class QuizQuestion(Base):
    __tablename__ = "quiz_questions"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    session_id: Mapped[str] = mapped_column(String, ForeignKey("quiz_sessions.id", ondelete="CASCADE"), nullable=False)
    question_index: Mapped[int] = mapped_column(Integer, nullable=False)
    question_text: Mapped[str] = mapped_column(Text, nullable=False)
    question_type: Mapped[str] = mapped_column(String(30), nullable=False)
    difficulty: Mapped[str] = mapped_column(String(10), nullable=False)
    options: Mapped[list | None] = mapped_column(JSON)
    correct_answer: Mapped[str] = mapped_column(Text, nullable=False)
    explanation: Mapped[str | None] = mapped_column(Text)
    source_page: Mapped[int | None] = mapped_column(Integer)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    session: Mapped["QuizSession"] = relationship("QuizSession", back_populates="questions")
    answer: Mapped["QuizAnswer | None"] = relationship("QuizAnswer", back_populates="question", uselist=False)


class QuizAnswer(Base):
    __tablename__ = "quiz_answers"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    session_id: Mapped[str] = mapped_column(String, ForeignKey("quiz_sessions.id", ondelete="CASCADE"), nullable=False)
    question_id: Mapped[str] = mapped_column(String, ForeignKey("quiz_questions.id", ondelete="CASCADE"), nullable=False)
    user_answer: Mapped[str | None] = mapped_column(Text)
    is_correct: Mapped[bool | None] = mapped_column(Boolean)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    session: Mapped["QuizSession"] = relationship("QuizSession", back_populates="answers")
    question: Mapped["QuizQuestion"] = relationship("QuizQuestion", back_populates="answer")


class QuizResult(Base):
    __tablename__ = "quiz_results"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    session_id: Mapped[str] = mapped_column(String, ForeignKey("quiz_sessions.id", ondelete="CASCADE"), nullable=False)
    score: Mapped[int] = mapped_column(Integer, nullable=False)
    total: Mapped[int] = mapped_column(Integer, nullable=False)
    percentage: Mapped[float] = mapped_column(Float, nullable=False)
    grade: Mapped[str | None] = mapped_column(String(5))
    weak_topics: Mapped[list | None] = mapped_column(JSON)
    recommendation: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    session: Mapped["QuizSession"] = relationship("QuizSession", back_populates="result")